File: src/main/java/com/baidu/algorithm/hashtable/_166_Fraction_to_Recurring_Decimal.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        sign = "-" if ((numerator ^ denominator) >> 31) else ""
        numerator, denominator = abs(numerator), abs(denominator)

        int_part, numerator = divmod(numerator, denominator)

        if numerator == 0:
            return "0" if not int_part else sign + str(int_part)

        return sign + str(int_part) + "." + str(self.helper(numerator, denominator))

    def helper(self, numerator, denominator):
        index = 0
        mp = {numerator: index}
        li = []
        while numerator:
            numerator *= 10
            li.append(str(numerator / denominator))

            numerator %= denominator
            if numerator in mp:
                begin = mp.get(numerator)
                li.insert(begin, '(')
                li.append(')')
                break

            index += 1
            mp[numerator] = index

        return "".join(li)

File: src/main/java/com/baidu/algorithm/hashtable/_166_Fraction_to_Recurring_Decimal.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.hashtable;

import java.util.HashMap;
import java.util.Map;

/**
 * _166_Fraction_to_Recurring_Decimal
 *
 * @author xuhaoran01
 */
public class _166_Fraction_to_Recurring_Decimal {
    private String helper(long n, long d) {
        int index = 0;

        Map<Long, Integer> map = new HashMap<>();
        map.put(n, index);

        StringBuilder sb = new StringBuilder();
        while (n > 0) {
            n *= 10;
            sb.append(n / d);
            n %= d;

            if (map.containsKey(n)) {
                int x = map.get(n);
                sb.insert(x, '(');
                sb.append(')');
                break;
            }

            index += 1;
            map.put(n, index);
        }

        return sb.toString();
    }

    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) {
            return "0";
        }

        long n = Math.abs((long) numerator);
        long d = Math.abs((long) denominator);
        String sign = ((numerator ^ denominator) & (1 << 31)) != 0 ? "-" : "";

        String left = String.valueOf(n / d);
        String right = (n % d != 0) ? helper(n % d, d) : "0";

        if ("0".equals(right)) {
            return sign + left;
        } else {
            return sign + left + "." + right;
        }
    }
}
